pkg/services/searchV2/readme.md

```diff
@@ -0,0 +1,98 @@
+Here is a proposal about a new package layout which:
+
+* will make different types of indexes work separately, each consuming `entity_events` individually in its own control goroutine, so less contention
+* will allow doing re-indexing with different periods for different types of indexes
+* will allow doing isolated backups, will simplify extending in custom way
+
+Entities:
+
+* `StandardSearchService` - wrapper for all known indexes. For now only has reference to `dashboardSearch`.
+* `dashboardSearch` - manages dashboard indexes for all organizations, every organization has its own `dashboardIndex`.
+* every `dashboardIndex` has its own `*bluge.Writer`
+
+```
+type StandardSearchService struct {
+  dashboardSearch *dashboardSearch
+}
+
+type dashboardSearch struct {
+  indexes map[int64]*dashboardIndex
+}
+
+type dashboardIndex struct {
+  writer *bluge.Writer
+}
+
+func (*dashboardSearch) run(ctx) error {
+  // Create initial indexes for known orgs.
+  // Start applying events, do full re-indexing, do backups etc.
+}
+
+func (*dashboardSearch) backup(ctx, currentEventID) error {
+  // write meta.json
+  // for each org do the backup using *bluge.Writer.
+  // done.
+}
+```
+
+Backup directory is configurable per index type. Inside backup directory structure may look like this:
+
+```
+dashboard
+   ├-- meta.json
+   ├── org1
+   │   ├── 0000000004a7.seg
+   │   ├── 0000000004a8.seg
+   │   ├── 0000000004a9.seg
+   │   └── 000000000893.snp
+   └── org2
+```
+
+`meta.json` contains:
+
+```
+{
+  "eventId": 12
+}
+```
+
+* `eventId` - the last event ID applied to the index in the backup. Since we make backups periodically we may need to apply some missing updates from `entity_event` table to catch up the state.
+
+As we don't have org id separation in `entity_event` table we manage indexes for all organizations in one goroutine. By different types of indexes are a separate consumers of `entity_event` table - so different types of indexes do not depend on each other at all.
+
+`IndexManager` is responsible controlling index lifecycle (actually this is our current `run` method), it accepts interface like:
+
+```
+type Index interface {
```

Comment:
Would it be maybe possible to have the `Index` / `IndexManager` per org to simplify its implementation?

Reply:
I suppose this is a bit problematic since we don't have incremental event ids in `entity_events` table with org resolution. Even if we had - in case of 300 orgs case I recently seen somewhere in issues this may mean 300 goroutines.

Reply:
What about having a top-level polling goroutine?

```
for {
  newEvents := getEvents(lastEventId)
  orgIdToEvents := partitionByOrgId(newEvents)
  for orgId, events := range orgIdToEvents {
      getIndexForOrg(orgId).applyUpdates(events)
   }

  sleep(5s)
}
```

Again, I am not sure if its going to help, but I think that as a general principle the fewer concepts we have in the core logic the better

Reply:
I think the problem here is that we need to have a single backup for all orgs - because last processed event ID is common for all organizations and we need some code that knows how to iterate over all currently indexed orgs and does snapshot to one dir with last processed event ID on top level.

But, maybe we can change interfaces a bit so that IndexManager does backups, but each individual indexes should implement a method like:

```
BackupTo(directory string) error
```

And IndexManager collects all backups from each required org.